### ISSUE TITLE
Some quesition about calculating KL divergence

### ISSUE THREAD

**Issue by LeoLiu1998** [wohlert/semi-supervised-pytorch#11]
I notice in the function:https://github.com/wohlert/semi-supervised-pytorch/blob/7cde4959468d271552febdcfed5e1cfae9857613/semi-supervised/models/vae.py#L200, you are calculating the KL divergence as in 
`qz - pz`

I am wondering, shouldn't the fomular should look something like 
`torch.exp(pz) * (qz - pz)`

according to the KL divergence definition

**Comment by wohlert** [comment]
The KL-divergence is specified as follows
![](https://wikimedia.org/api/rest_v1/media/math/render/svg/4958785faae58310ca5ab69de1310e3aafd12b32)
You will notice that we do the same except we switch the name of q and p as we prefer the mode-seeking behaviour of "reverse KL". You will see that the original [Gumbel-softmax implementation in Tensorflow](https://github.com/ericjang/gumbel-softmax/blob/master/Categorical%20VAE.ipynb) does the same.

**Comment by LeoLiu1998** [comment]
Oh, I am sorry, I link the wrong line. 

This is what I really mean: https://github.com/wohlert/semi-supervised-pytorch/blob/7cde4959468d271552febdcfed5e1cfae9857613/semi-supervised/models/vae.py#L118

**Comment by wohlert** [comment]
Ah yes, as I show in the documentation:
```
 KL(q||p) = -∫ q(z) log [ p(z) / q(z) ]
                 = -E[log p(z) - log q(z)]
```
This is another way to calculate it if we just have samples `z` drawn from the q-distribution. This is also known as a Monte-Carlo approximation to the expectation.

**Comment by LeoLiu1998** [comment]
So the method is assming sample size =  1? or it depends on how we use the method?

**Comment by wohlert** [comment]
The number of samples is specified in the inference method in [SVI](https://github.com/wohlert/semi-supervised-pytorch/blob/master/semi-supervised/inference/variational.py#L15)

**Comment by YannDubs** [comment]
Hi @wohlert , thank you for the clean code and nice documentation. 
I was wondering why you used MC approximation of the KL divergence instead of the closed form solution when both distributions are Gaussians (for example p_z and q_z) ? 

Is it simply to have a more general code, or is there a deeper reason?

Cheers

**Comment by wohlert** [comment]
At the time I was experimenting with a lot of different distributions and `torch.distributions` wasn't implemented yet, so this was the best approach. For now, I would recommend either using the closed form solution or `torch.distributions`.
